### PR TITLE
Fix random generation, to not always retrieve 0 after seed()

### LINKED ISSUE #17664: randi() returns 0 for FIRST number after seed()
CODE:
```
seed(0)
print(randi())
print(randi())
seed(1)
print(randi())
print(randi())
```

OUTPUT (Godot 2.1.4):
```
719435623
1221583151
16807
282475249
```

OUTPUT (Godot 3.0.2):
```
Godot 3.0.2
0
1613493245
0
3114030964
```

If `randomize()` is used, the same issue doesn't seem to occur.

Is this expected behaviour?

### PR DESCRIPTION
**Pull request by Chaosus**
Fix #17664

### CODE REVIEW

**Comment by Chaosus** [comment]
Im not sure that it will break compatibility, pushed to 3.1 for now

File: thirdparty/misc/pcg.cpp [hunk 1]
```diff
@@ -3,13 +3,16 @@
...
-uint32_t pcg32_random_r(pcg32_random_t* rng)
-{
-    uint64_t oldstate = rng->state;
-    // Advance internal state
-    rng->state = oldstate * 6364136223846793005ULL + (rng->inc|1);
-    // Calculate output function (XSH RR), uses old state for max ILP
-    uint32_t xorshifted = ((oldstate >> 18u) ^ oldstate) >> 27u;
-    uint32_t rot = oldstate >> 59u;
-    return (xorshifted >> rot) | (xorshifted << ((-rot) & 31));
+void pcg32_random_init(pcg32_random_t *rng) {
```
Comment (akien-mga) [inline thirdparty/misc/pcg.cpp]: This is thirdparty code, so I would prefer if we don't change it.

If we do need to change it, we should at least not change the code style of the thirdparty code, as it generates a bigger diff than necessary. And the changes we made should be clearly identified with `// -- GODOT -- start` and `// -- GODOT -- end` comments.

Comment (Chaosus) [reply thirdparty/misc/pcg.cpp]: Yeah, it was my fault, I didnt pay attention to thirdparty tag. Well I guess generating one number leads to the same result(with a little performance overhead).

**Comment by Chaosus** [comment]
@akien-mga I fixed it

File: core/math/random_pcg.h [hunk 2]
```diff
@@ -45,7 +45,10 @@ class RandomPCG {
 
 	RandomPCG(uint64_t seed = DEFAULT_SEED, uint64_t inc = PCG_DEFAULT_INC_64);
 
-	_FORCE_INLINE_ void seed(uint64_t seed) { pcg.state = seed; }
+	_FORCE_INLINE_ void seed(uint64_t seed) {
+		pcg.state = seed;
+		pcg32_random_r(&pcg);
```
Comment (akien-mga) [inline core/math/random_pcg.h]: Can you add a comment explaining why we have to do this when seeding?

Comment (Chaosus) [reply core/math/random_pcg.h]: Done

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Note that it does break compat as a given seed in 3.0.x will no longer produce the same results. So if people used the seed to have deterministic RNG (e.g. for a given procedurally generated level, etc.), it won't upgrade smoothly from 3.0.x to 3.1.

It's still an OK change for 3.1, but the compat change needs to be documented accordingly in the changelog (cc @Calinou).

**Comment by MidZik** [comment]
I'm going to be doing some testing to verify this, but doesn't this break saving and restoring the RNG state? If you save the seed and then restore it, the RNG won't actually be in the same state it was in before.

**Comment by akien-mga** [comment]
Yes, that's what I mentioned above:

> Note that it does break compat as a given seed in 3.0.x will no longer produce the same results. So if people used the seed to have deterministic RNG (e.g. for a given procedurally generated level, etc.), it won't upgrade smoothly from 3.0.x to 3.1.

**Comment by MidZik** [comment]
I don't mean between versions, I mean within the same version.

My RNG is in some state (seed x). I save this state to a file.

Later I restore that state. The RNG is now in state (seed x + 1). It will not behave the same way as when it was saved.

**Comment by akien-mga** [comment]
Hm, that's right. That would be pretty bad. If you can test it and confirm, we should likely revert this change.

**Comment by MidZik** [comment]
Yeah, it seems one project I had before that relied on saving/loading RNG state no longer works properly. I can create a minimum repro project if you want, but I think the issue is fairly clear.

I agree that this should be reverted as it is now, although fixing this might not be too difficult either. I think you can just keep track of both the previous and internal seed, and the public `seed` property will always point to the previous seed.

**Comment by MidZik** [comment]
I'll give fixing this a shot before you decide to revert it.

**Comment by Chaosus** [comment]
Yeah, feel a bit sorry about that, thankfully 3.1 is not yet released..

**Comment by Chaosus** [comment]
However if it will return the correct seed as #26737 proposed why this need to be reverted ?

**Comment by MidZik** [comment]
If my fix is accepted then it won't be reverted. 🙂

**Pull request merged**